From 613a087bb82e1243163934b9c681119f039efcd7 Mon Sep 17 00:00:00 2001
From: Dinolek <git@dinolek.me>
Date: Tue, 1 Jul 2025 13:40:00 +0200
Subject: [PATCH 087/115] Input: novatek-nvt-ts - add support for panel
 follower

Touchscreen driver needs to be initialized after the DSI part is done.
Touch controllers without persistent memory need to upload firmware when
enabling the panel.

Signed-off-by: Dinolek <git@dinolek.me>
---
 drivers/input/touchscreen/Kconfig             |  1 +
 .../input/touchscreen/novatek-nvt-ts-core.c   | 77 +++++++++++++++++--
 drivers/input/touchscreen/novatek-nvt-ts.h    |  5 ++
 3 files changed, 75 insertions(+), 8 deletions(-)

diff --git a/drivers/input/touchscreen/Kconfig b/drivers/input/touchscreen/Kconfig
index 6788aa157..30f6ab6f3 100644
--- a/drivers/input/touchscreen/Kconfig
+++ b/drivers/input/touchscreen/Kconfig
@@ -694,6 +694,7 @@ config TOUCHSCREEN_MTOUCH
 
 config TOUCHSCREEN_NOVATEK_NVT_TS_CORE
 	tristate
+	depends on DRM || !DRM
 
 config TOUCHSCREEN_NOVATEK_NVT_TS_I2C
 	tristate "Novatek I2C touchscreen support"
diff --git a/drivers/input/touchscreen/novatek-nvt-ts-core.c b/drivers/input/touchscreen/novatek-nvt-ts-core.c
index 3ce9bf169..565a28a61 100644
--- a/drivers/input/touchscreen/novatek-nvt-ts-core.c
+++ b/drivers/input/touchscreen/novatek-nvt-ts-core.c
@@ -131,10 +131,19 @@ static void nvt_ts_stop(struct input_dev *dev)
 	regulator_bulk_disable(ARRAY_SIZE(data->regulators), data->regulators);
 }
 
-static int nvt_ts_suspend(struct device *dev)
+static void nvt_ts_panel_prepare_work(struct work_struct *work)
 {
-	const struct nvt_ts_data *data = dev_get_drvdata(dev);
+	struct nvt_ts_data *data = container_of(work, struct nvt_ts_data, panel_follower_prepare_work);
 
+	int ret = nvt_ts_start(data->input);
+	if (!ret) {
+		WRITE_ONCE(data->prepare_work_finished, true);
+		smp_wmb();
+	}
+}
+
+static int nvt_ts_input_stop(const struct nvt_ts_data *data)
+{
 	guard(mutex)(&data->input->mutex);
 
 	if (input_device_enabled(data->input))
@@ -143,18 +152,63 @@ static int nvt_ts_suspend(struct device *dev)
 	return 0;
 }
 
+static int nvt_ts_suspend(struct device *dev)
+{
+	struct nvt_ts_data *data = dev_get_drvdata(dev);
+
+	if (drm_is_panel_follower(dev))
+		return 0;
+
+	cancel_work_sync(&data->panel_follower_prepare_work);
+
+	smp_rmb();
+	if (!READ_ONCE(data->prepare_work_finished))
+		return 0;
+
+	return nvt_ts_input_stop(data);
+}
+
 static int nvt_ts_resume(struct device *dev)
 {
-	const struct nvt_ts_data *data = dev_get_drvdata(dev);
+	struct nvt_ts_data *data = dev_get_drvdata(dev);
 
-	guard(mutex)(&data->input->mutex);
+	if (drm_is_panel_follower(dev))
+		return 0;
 
-	if (input_device_enabled(data->input))
-		nvt_ts_start(data->input);
+	WRITE_ONCE(data->prepare_work_finished, false);
+	schedule_work(&data->panel_follower_prepare_work);
 
 	return 0;
 }
 
+static int panel_prepared(struct drm_panel_follower *follower)
+{
+	struct nvt_ts_data *data = container_of(follower, struct nvt_ts_data, panel_follower);
+
+	WRITE_ONCE(data->prepare_work_finished, false);
+	schedule_work(&data->panel_follower_prepare_work);
+
+	return 0;
+}
+
+static int panel_unpreparing(struct drm_panel_follower *follower)
+{
+	struct nvt_ts_data *data = container_of(follower, struct nvt_ts_data, panel_follower);
+
+	cancel_work_sync(&data->panel_follower_prepare_work);
+
+	smp_rmb();
+	if (!READ_ONCE(data->prepare_work_finished))
+		return 0;
+
+	return nvt_ts_input_stop(data);
+}
+
+static struct drm_panel_follower_funcs nvt_ts_panel_follower_funcs = {
+	.panel_prepared = panel_prepared,
+	.panel_unpreparing = panel_unpreparing,
+};
+
 EXPORT_GPL_SIMPLE_DEV_PM_OPS(nvt_ts_pm_ops, nvt_ts_suspend, nvt_ts_resume);
 
 int nvt_ts_probe(struct device *dev, int irq, struct regmap *regmap,
@@ -253,8 +307,6 @@ int nvt_ts_probe(struct device *dev, int irq, struct regmap *regmap,
 
 	input->name = dev_name(dev);
 	input->id = *id;
-	input->open = nvt_ts_start;
-	input->close = nvt_ts_stop;
 
 	input_set_abs_params(input, ABS_MT_POSITION_X, 0, width - 1, 0, 0);
 	input_set_abs_params(input, ABS_MT_POSITION_Y, 0, height - 1, 0, 0);
@@ -275,6 +327,15 @@ int nvt_ts_probe(struct device *dev, int irq, struct regmap *regmap,
 	if (error)
 		return dev_err_probe(dev, error, "failed to request irq\n");
 
+	if (drm_is_panel_follower(dev)) {
+		INIT_WORK(&data->panel_follower_prepare_work, nvt_ts_panel_prepare_work);
+		data->panel_follower.funcs = &nvt_ts_panel_follower_funcs;
+		devm_drm_panel_add_follower(dev, &data->panel_follower);
+	} else {
+		input->open = nvt_ts_start;
+		input->close = nvt_ts_stop;
+	}
+
 	error = input_register_device(input);
 	if (error)
 		return dev_err_probe(dev, error, "failed to register input device\n");
diff --git a/drivers/input/touchscreen/novatek-nvt-ts.h b/drivers/input/touchscreen/novatek-nvt-ts.h
index 4fb2f96ca..015c6d9d4 100644
--- a/drivers/input/touchscreen/novatek-nvt-ts.h
+++ b/drivers/input/touchscreen/novatek-nvt-ts.h
@@ -2,6 +2,8 @@
 #ifndef __NOVATEK_NVT_TS_H_
 #define __NOVATEK_NVT_TS_H_
 
+#include <drm/drm_panel.h>
+
 #include <linux/firmware.h>
 #include <linux/input.h>
 #include <linux/input/touchscreen.h>
@@ -71,6 +73,9 @@ struct nvt_ts_data {
 	int irq;
 	nvt_ts_upload_firmware_impl upload_firmware;
 	const struct nvt_ts_memory_map *memory_map;
+	struct drm_panel_follower panel_follower;
+	struct work_struct panel_follower_prepare_work;
+	bool prepare_work_finished;
 	const char *fw_name;
 	u8 buf[NVT_TS_TOUCH_SIZE * NVT_TS_MAX_TOUCHES];
 };
-- 
2.56.0

